timely/src/progress/subgraph.rs

```diff
@@ -530,26 +530,37 @@ where
         assert_eq!(self.children[0].outputs, self.inputs());
         assert_eq!(self.children[0].inputs, self.outputs());
 
-        let mut internal_summary = Vec::with_capacity(self.inputs());
-        for summary in self.scope_summary.iter() {
-            let scope_summary = summary.iter()
-                .map(|output| {
-                    output.elements()
-                        .iter()
-                        .cloned()
-                        .fold(
-                            Antichain::with_capacity(output.elements().len()),
-                            |mut antichain, path_summary| {
-                                antichain.insert(TInner::summarize(path_summary));
-                                antichain
-                            },
-                        )
-                })
-                .collect();
-
-            internal_summary.push(scope_summary);
+        // Note that we need to have `self.inputs()` elements in the summary
+        // with each element containing `self.outputs()` antichains regardless
+        // of how long `self.scope_summary` is
+        let mut internal_summary: Vec<Vec<_>> = (0..self.inputs())
+            .map(|_| (0..self.outputs()).map(|_| Antichain::new()).collect())
+            .collect();
+
+        for (summary, internal_summary) in
+            self.scope_summary.iter().zip(internal_summary.iter_mut())
+        {
+            internal_summary.reserve(summary.len());
+
+            for (output, antichain) in summary.iter().zip(internal_summary) {
+                antichain.reserve(output.elements().len());
+
+                for path_summary in output.elements() {
+                    antichain.insert(TInner::summarize(path_summary.clone()));
+                }
+            }
         }
 
+        debug_assert_eq!(
+            internal_summary.len(),
+            self.inputs(),
+            "the internal summary should have as many elements as there are inputs",
+        );
+        debug_assert!(
+            internal_summary.iter().all(|summary| summary.len() == self.outputs()),
+            "each element of the internal summary should have as many elements as there are outputs",
+        );
```

Comment:
These are good, but tbh we should probably steal them for the assert that was failing, so that they apply to all operators rather than just subgraphs.

Reply:
I'm not sure what you mean by that, you mean to move the assertions to another place?

Reply:
This is fine, sorry. I think it can be statically verified at this point (with the explicit builder). I was thinking that `PerOperatorState::new()` is the place that one might want to re-verify, but there is already the assert there (without the nice error statements).